flask/app/utils.py

```diff
@@ -154,6 +165,73 @@ def get_minio_admin() -> MinioAdmin:
     )
 
 
+def query_results_to_csv(results: List[dict or dataclass]):
+    """ take a list of models and convert to csv """
+    is_dict = isinstance(results[0], dict)
+
+    if not is_dict:
+        results = [asdict(result) for result in results]
+
+    colnames = results[0].keys()
+
+    csv_data = StringIO()
+    writer = DictWriter(
+        csv_data,
+        fieldnames=colnames,
+        quoting=QUOTE_MINIMAL,
+    )
+    writer.writeheader()
+    for row in results:
+        if isinstance(row, Model):
+            row = asdict(row)
+        writer.writerow(row)
+
+    return csv_data.getvalue()
+
+
+def paginated_response(results: List[Any], page: int, total: int, limit: int = None):
+    return jsonify(
+        {
+            "data": results,
+            "page": page if limit else 0,
+            "total_count": total,
+        }
+    )
+
+
+def csv_response(
+    results: List[Dict[str, Any]] or List[Model],
+    filename: str = "report",
+    colnames: List[str] = None,
+):
+
+    if colnames:
+        results = filter_keys_and_reorder(results, colnames)
+
+    """ create a csv HTTP response from a list of query results or mapping """
+    csv = query_results_to_csv(results)
+
+    return send_file(
+        BytesIO(csv.encode("utf-8")),
+        "text/csv",
+        as_attachment=True,
+        attachment_filename=filename,
+    )
+
+
+def filter_keys_and_reorder(data: List[Dict[str, Any]] or List[Model], keys: List[str]):
+    """ filter list item mappings and key order according to list of keys """
+    return [{key: row[key] for key in keys} for row in data]
+
+
+def expects_json(req: Request):
+    return req.headers.get("Accept") in ["application/json", "*/*", None]
+
+
+def expects_csv(req: Request):
+    return req.headers.get("Accept") == "text/csv"
```

Comment:
I think `request.accept_mimetypes` was fine because Flask took care of figuring out that `*/*` means you accept `application/json`.

Reply:
I did it this way b/c the integration test requests don't have the Accept header set, so the existing json-expecting tests for analyses, datasets, participants will fail unless it's added (hence the `None` option). I can add the header to the tests and restore the mimetypes check (or do something like `if not request.accept_mimetypes or 'application/json' in request.accept_mimetypes`) if that makes more sense.

Reply:
I see. I think that's more a problem with how we do tests, since the `request.accept_mimetypes` object takes care of a lot of things for us with its convenient syntax.

Reply:
Sounds good. I'll add the headers to the tests.

Reply:
I think you might be able to cheese it with what the proposed `if not request.accept_mimetypes or 'application/json' in request.accept_mimetypes`